# <h1 style="text-align: center; margin: 0"> MASA-CUDAlign_MAPA </h1>

<p>Sequence alignment plays a crucial role in molecular biology and bioinformatics, serv-
ing as a fundamental technique in understanding DNA, RNA, and proteins. CUDAlign
stands out by utilizing GPUs to efficiently achieve optimal alignments. In CUDAlign
2.1, the pruning process was introduced, involving the removal of the computation of the
blocks from dynamic programming matrices that do not contribute to optimal alignment,
thus accelerating the alignment algorithm. The APA module (Agile Pruning Alignment)
was designed to enhance the block pruning process, generating a heuristic score used
as the initial score for CUDAlign, aiming to expand the pruning area. However, the
synchronous execution of APA introduces overhead in the overall execution time. This
work presents the design of MAPA (Asynchronous Pruning Acceleration Module) with
the goal of eliminating this overhead. The results obtained demonstrate a considerable
performance improvement compared to the APA module.</p>

<h3>How to execute MAPA</h3>

```
git clone https://github.com/trajano7/masa_cudalign_MAPA.git
cd masa_cudalign_MAPA/cudalign
./configure --with-cuda=/usr/lib/cuda --with-nvcc=/usr/lib/nvidia-cuda-toolkit/bin/
make
cd ..
g++ -std=c++17 mapa.cpp -o mapa
./mapa [execution options] sequenciaA.fasta sequenciaB.fasta
```

<p>The execution options are:</p>
<ul>
<li><em>-h</em> or <em>--help</em>: show the help message</li>
<li><em>-ns</em> or <em>--noscore</em>: execution just of the CUDAlign, without the use the heuristic score</li>
<li><em>-t</em> or <em>--trim</em>: execution with a heuristic score generated by trimmed sequences</li>
<li><em>-f</em> or <em>--full</em>: execution with a heuristic score generated by full sequences</li>
</ul>


                   